protocols/imap/src/main/scala/org/apache/james/imap/decode/parser/SearchCommandParser.scala

```diff
@@ -0,0 +1,683 @@
+      case _ =>
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+  }
+
+  @throws[DecodingException]
+  private def old(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("old", alreadyConsumed)(request)
+    try { // Check if its OLDER keyword
+      nextIs(request, 'e')
+      older(request, alreadyConsumed + 1)
+    } catch {
+      case _: DecodingException =>
+        Old()
+    }
+  }
+
+  @throws[DecodingException]
+  private def n(session: ImapSession, request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey =
+    consumeAndCap(request) match {
+      case 'E' =>
+        newOperator(request, alreadyConsumed + 1)
+      case 'O' =>
+        not(session, request, charset, alreadyConsumed + 1)
+      case _ =>
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+  }
+
+  @throws[DecodingException]
+  private def f(request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey =
+    consumeAndCap(request) match {
+      case 'L' =>
+        flagged(request, alreadyConsumed + 1)
+      case 'R' =>
+        from(request, charset, alreadyConsumed + 1)
+      case _ =>
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+  }
+
+  @throws[DecodingException]
+  private def d(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey =
+    consumeAndCap(request) match {
+      case 'E' =>
+        deleted(request, alreadyConsumed + 1)
+      case 'R' =>
+        draft(request, alreadyConsumed + 1)
+      case _ =>
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+  }
+
+  @throws[DecodingException]
+  private def keyword(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("keyword ", alreadyConsumed)(request)
+    Keyword(request.atom)
+  }
+
+  @throws[DecodingException]
+  private def unkeyword(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("unkeyword ", alreadyConsumed)(request)
+    UnKeyword(request.atom)
+  }
+
+  @throws[DecodingException]
+  private def header(request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("header ", alreadyConsumed)(request)
+    val field = request.astring(charset)
+    nextIsSpace(request)
+    val value = request.astring(charset)
+    Header(field, value)
+  }
+
+  @throws[DecodingException]
+  private def larger(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("larger ", alreadyConsumed)(request)
+    Larger(request.number)
+  }
+
+  @throws[DecodingException]
+  private def smaller(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("smaller ", alreadyConsumed)(request)
```

Comment:
While I do agree with the change to carry over consumed characters in order to assert equality on the full token,

this is a behavioural change in the middle of a "migrate to scala" commit.

How can I review such a thing?

Reply:
Sorry, I don't understand what behavior changed, could you elaborate?

Reply:
Before that, consumed characters were not part of the consumed values.

Given a consumed 's' here the expected characters are 'maller'.

Though I agree with the change.

Reply:
I'm not getting your point.

Here, we will just match `maller ` as before, it's just nicer to read because I put the full token (and skip the part already matcher thanks to `alreadyConsumed`)

Reply:
Well it's nicer to read but it is not a simple java -> scala migration as the arguments of these method changes.

Making sure "that it behaves as before" is thus harder.

Separating simple scala migration from these readability improvment could have been helping.

Reply:
Never said it's simple java -> scala migration.
It's what refactoring is for: make it nicer while changing nothing about how it behaves.
I can understand it's could be hard to review but the non-refactored version was worse (: